libbpf-tools/execsnoop.bpf.c

```diff
@@ -0,0 +1,129 @@
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include "execsnoop.h"
+
+// null isn't defined
+#define NULL 0
+
+const volatile bool ignore_failed = true;
+const volatile uid_t targ_uid = INVALID_UID;
+const volatile int max_args = DEFAULT_MAXARGS;
+
+static const struct event empty_event = {};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, pid_t);
+	__type(value, struct event);
+} execs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline bool valid_uid(uid_t uid) {
+	return uid != INVALID_UID;
+}
+
+SEC("tracepoint/syscalls/sys_enter_execve")
+int tracepoint__syscalls__sys_enter_execve(struct trace_event_raw_sys_enter* ctx)
+{
+	u64 id;
+	pid_t pid, tgid;
+	int ret;
+	struct event *event;
+	uid_t uid = (u32)bpf_get_current_uid_gid();
+	if (valid_uid(targ_uid) && targ_uid != uid)
+		return 0;
+
+	id = bpf_get_current_pid_tgid();
+	pid = (pid_t)id;
+	tgid = id >> 32;
+	if (bpf_map_update_elem(&execs, &pid, &empty_event, BPF_NOEXIST))
+		return 0;
+
+	event = bpf_map_lookup_elem(&execs, &pid);
+	if (!event)
+		return 0;
+
+	event->pid = pid;
+	event->tgid = tgid;
+	event->uid = uid;
+	event->more_args = false;
+	event->args_count = 0;
+	event->args_size = 0;
+
+	ret = bpf_probe_read_str(event->args, ARGSIZE, (const char*)ctx->args[0]);
+	if (ret > 0 && ret <= ARGSIZE) {
```

Comment:
it's simpler to declare ret as unsigned integer, and then this if will turn into:
```
if (ret <= ARGSIZE) {
    event->args_size += ret;
}
```

Reply:
IMO, it's very hacky, also it doesn't check that ret isn't equal to `0` (this for some reason had issues with verifier).

Reply:
It's a very common practice in BPF programs, it can generate simpler and more performance BPF code as well, as well as the one that's easier to verify for verifier. So I'd say still do it. You can leave comment that negative return will turn into a large unsigned ret value.

bpf_probe_read_str() can't return zero if you pass size >0. What kind of problem does it cause for verifier?

Reply:
Odd, now it works.